Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -550,28 +550,62 @@ PhysicalParticleContainer::AddGaussianBeam (
         for (long i = 0; i < npart; ++i) {
 #if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
             const Real weight = q_tot/(npart*charge);
-            const Real x = amrex::RandomNormal(x_m, x_rms);
-            const Real y = amrex::RandomNormal(y_m, y_rms);
-            const Real z = amrex::RandomNormal(z_m, z_rms);
+            Real x = amrex::RandomNormal(x_m, x_rms);
+            Real y = amrex::RandomNormal(y_m, y_rms);
+            Real z = amrex::RandomNormal(z_m, z_rms);
 #elif defined(WARPX_DIM_XZ)
             const Real weight = q_tot/(npart*charge*y_rms);
-            const Real x = amrex::RandomNormal(x_m, x_rms);
+            Real x = amrex::RandomNormal(x_m, x_rms);
             constexpr Real y = 0._prt;
-            const Real z = amrex::RandomNormal(z_m, z_rms);
+            Real z = amrex::RandomNormal(z_m, z_rms);
 #elif defined(WARPX_DIM_1D_Z)
             const Real weight = q_tot/(npart*charge*x_rms*y_rms);
             constexpr Real x = 0._prt;
             constexpr Real y = 0._prt;
-            const Real z = amrex::RandomNormal(z_m, z_rms);
+            Real z = amrex::RandomNormal(z_m, z_rms);
 #endif
             if (plasma_injector.insideBounds(x, y, z)  &&
                 std::abs( x - x_m ) <= x_cut * x_rms     &&
                 std::abs( y - y_m ) <= y_cut * y_rms     &&
                 std::abs( z - z_m ) <= z_cut * z_rms   ) {
                 XDim3 u = plasma_injector.getMomentum(x, y, z);
+
+            if (plasma_injector.do_focusing){
+                XDim3 u_bulk = plasma_injector.getInjectorMomentumHost()->getBulkMomentum(x,y,z);
+                Real u_bulk_norm = std::sqrt( u_bulk.x*u_bulk.x+u_bulk.y*u_bulk.y+u_bulk.z*u_bulk.z );
+                Real gamma_bulk = std::sqrt(1._rt + u_bulk.x*u_bulk.x+u_bulk.y*u_bulk.y+u_bulk.z*u_bulk.z );
+
+                Real x_f = x_m + focal_distance * u_bulk.x/u_bulk_norm;
+                Real y_f = y_m + focal_distance * u_bulk.y/u_bulk_norm;
+                Real z_f = z_m + focal_distance * u_bulk.z/u_bulk_norm;
+                Real gamma = std::sqrt( 1._rt + (u.x*u.x+u.y*u.y+u.z*u.z) );
+
+                Real v_x = u.x / gamma * PhysConst::c;
+                Real v_y = u.y / gamma * PhysConst::c;
+                Real v_z = u.z / gamma * PhysConst::c;
+
+                Real d = std::sqrt( (x_f-x)*(x_f-x) + (y_f-y)*(y_f-y) + (z_f-z)*(z_f-z) );
+                Real t = d / std::sqrt(v_x*v_x+v_y*v_y+v_z*v_z);
+
+                amrex::Real v_bulk_z = u_bulk.z / gamma_bulk * PhysConst::c;
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+                Real v_bulk_x = u_bulk.x / gamma_bulk * PhysConst::c;
+                Real v_bulk_y = u_bulk.y / gamma_bulk * PhysConst::c;
+                x = x - (v_x - v_bulk_x) * t;
+                y = y - (v_y - v_bulk_y) * t;
+                z = z - (v_z - v_bulk_z) * t;
```

Comment:
Here also, I think that a slightly more correct expression would be
```suggestion
                Real v_dot_n = v_x * n_x + v_y * n_y + v_z * n_z;
                x = x - (v_x - v_dot_n*n_x) * t;
                y = y - (v_y - v_dot_n*n_y) * t;
                z = z - (v_z - v_dot_n*n_z) * t;
```

Reply:
Note that, in this case, you don't need to compute `gamma_bulk`.